File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
  def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
    stack = []
    next_greater_element = dict()

    for num in nums2[::-1]:
      while stack and stack[-1] <= num:
        stack.pop()
      next_greater_element[num] = stack[-1] if stack else -1
      stack.append(num)
    
    return [next_greater_element[num] for num in nums1]

File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int[] result = new int[nums1.length];
        Arrays.fill(result, -1);
        Deque<Integer> stack = new ArrayDeque<>();
        Map<Integer, Integer> indexOf = new HashMap<>(nums1.length);
        for (int i = 0; i < nums1.length; i++) {
            indexOf.put(nums1[i], i);
        }

        for (int i = 0; i < nums2.length; i++) {
            while (!stack.isEmpty() && nums2[i] > stack.peek()) {
                result[indexOf.get(stack.pop())] = nums2[i];
            }

            if (indexOf.containsKey(nums2[i])) {
                stack.push(nums2[i]);
            }
        }
        
        return result;
    }
}